lib/LLVMIRCodeGen/BundleSaver.cpp

```diff
@@ -190,7 +190,73 @@ static void serializeBinaryToText(llvm::StringRef binFileName,
 }
 
 BundleSaver::BundleSaver(const IRFunction *F, const LLVMBackend &llvmBackend)
-    : F_(F), irgen_(llvmBackend.createIRGen(F_, allocationsInfo_)) {}
+    : irgen_(llvmBackend.createIRGen(F, allocationsInfo_)) {
+  setIRFunction("", F);
+}
+
+BundleSaver::BundleSaver(const LLVMBackend &llvmBackend,
+                         llvm::StringRef outputDir, llvm::StringRef bundleName)
+    : irgen_(llvmBackend.createIRGen(nullptr, allocationsInfo_)),
+      outputDir_(outputDir), bundleName_(bundleName) {
+  llvm::SmallVector<std::string, 8> targetFeatures(llvmTargetFeatures.begin(),
+                                                   llvmTargetFeatures.end());
+  irgen_->setBundleName(bundleName);
+  irgen_->setOutputDir(outputDir);
+  irgen_->initTargetMachine(llvmBackend.getTarget(), llvmBackend.getArch(),
+                            llvmBackend.getCPU(), targetFeatures,
+                            llvmBackend.getBundleCodeModel(),
+                            llvmBackend.getRelocModel());
+  irgen_->initCodeGen();
+}
+
+void BundleSaver::setIRFunction(llvm::StringRef mainEntryName,
+                                const IRFunction *F) {
+  irgen_->setIRFunction(F);
+  if (F) {
+    savedIRFunctions_.push_back(SavedIRFunction{mainEntryName, F});
+  }
+}
+
+bool BundleSaver::WeightAddrComparator::
+operator()(const WeightInfo &LHS, const WeightInfo &RHS) const {
+  auto lhsAddr =
+      bundleSaver_.allocationsInfo_.allocatedAddress_.lookup(LHS.first);
+  auto rhsAddr =
+      bundleSaver_.allocationsInfo_.allocatedAddress_.lookup(RHS.first);
+  return lhsAddr < rhsAddr;
+}
+
+std::set<BundleSaver::WeightInfo, BundleSaver::WeightAddrComparator>
```

Comment:
not llvm's SetVector? :)

Reply:
Does `llvm::SetVector` support custom comparators?

Reply:
errm good point: it is clunkier with LLVM: would probably need to do that via explicitly via overriding the default hash value / implement their hashing API. forget about this nit - i withdraw my comment :)